common/reactive/src/main/java/io/helidon/common/reactive/BufferedEmittingPublisher.java

```diff
@@ -193,28 +332,54 @@ public boolean isUnbounded() {
      * @return true if demand is higher than 0
      */
     public boolean hasRequests() {
-        return this.emitter.hasRequests();
+        return requested.get() > emitted;
     }
 
     /**
      * Check if publisher sent {@code onComplete} signal downstream.
      * Returns {@code true} right after calling {@link BufferedEmittingPublisher#completeNow()}
+     * (with a caveat)
      * but after calling {@link BufferedEmittingPublisher#complete()} returns
      * {@code false} until whole buffer has been drained.
+     * <p>
+     * The caveat is that completeNow() does not guarantee that the onComplete signal is sent
+     * before returning from completeNow() - it is only guaranteed to be sent as soon as it can be done.
      *
      * @return true if so
      */
     public boolean isCompleted() {
-        return this.state.get() == State.COMPLETED;
+        // The caveat above means only that the current implementation guarantees onComplete is sent
+        // before completeNow returns in single-threaded cases. When concurrent emit() or request()
+        // race against completeNow, completeNow may return without entering drain() - but the concurrent
+        // calls guarantee onComplete will be called as soon as they observe the buffer is empty.
+        //
+        // We don't want to say this in the public documentation as this is implementation detail.
+        //
+        // A subtle logical statement: if onError or onComplete has been signalled to downstream,
+        //   isCompleted is true. But it is also true if cancel() precluded sending the signal
+        //   to downstream.
+        //
+        //   The current implementation is: isCompleted is true if and only if no more onNext signals
+        //   will be sent to downstream and no cancellation was requested.
+        //
+        // assert: once isCompleted becomes true, it stays true
+        // question: what should it be, if complete() was called, but not onSubscribe()?
+        return buffer.isEmpty() && state.get() > 1;
```

Comment:
The conditions must be checked in the opposite order: `state.get() > 1` first, then `buffer.isEmpty()`.

Reply:
awesome!